### PR TITLE
proj-data: init at 1.16.0

### PR DESCRIPTION
**Pull request by imincik**
## Description of changes

* add initial version of `proj-data` package containing datum grid resource files for `proj`
package
* add optional support for installing `proj-data` grid resource files in `proj` data directory

AFAIK, installation of `proj-data` resource files in `proj` directory is the only user-friendly way of enabling them for all programs using `proj`. Another alternative methods are:

* setting `PROJ_DATA` variable to `${proj}/share/proj:${proj-data}` (see `proj-data` package tests)
* adding them to user home directories

which can be very tricky for other applications using `proj` in containers or multi-user environments and can be error-prone.
For more information, see: https://proj.org/en/9.3/resource_files.html

Installation of `proj-data` files to `proj` is optional, because `proj-data` repository size is large (can reach 1GB of size soon) and it avoids downloading it when building `proj` package version where `proj-data` is not needed. 

### Example usage
* Example of building proj package containing `proj-data` grids provided by `nz_linz`:
```
$ nix-build -E 'with import ./. {}; proj.override { proj-data = proj-data.override { gridPackages = ["nz_linz"]; };  withProjData = true; }'

ls ./result/share/proj
CH                               nz_linz_nzgd2000-20130801.json           nz_linz_nzgd2000-ka20161114-grid02.tif   nz_linz_nzgd2000-ndm-grid01.tif
deformation_model.schema.json    nz_linz_nzgd2000-20140201.json           nz_linz_nzgd2000-ka20161114-grid03.tif   nz_linz_nzgd2000-ndm-grid02.tif
GL27                             nz_linz_nzgd2000-20150101.json           nz_linz_nzgd2000-ka20161114-grid04.tif   nz_linz_nzgd2000-si20030821-grid01.tif
ITRF2000                         nz_linz_nzgd2000-20160701.json           nz_linz_nzgd2000-ka20161114-grid05.tif   nz_linz_nzgd2kgrid0005.tif
ITRF2008                         nz_linz_nzgd2000-20171201.json           nz_linz_nzgd2000-ka20161114-grid06.tif   nz_linz_nzgeoid2009.tif
ITRF2014                         nz_linz_nzgd2000-20180701.json           nz_linz_nzgd2000-ka20161114-grid07.tif   nz_linz_nzgeoid2016.tif
nad27                            nz_linz_nzgd2000-c120100904-grid01.tif   nz_linz_nzgd2000-ka20161114-grid08.tif   nz_linz_ontpht1964-nzvd2016.tif
nad83                            nz_linz_nzgd2000-c220110222-grid01.tif   nz_linz_nzgd2000-ka20161114-grid09.tif   nz_linz_README.txt
nad.lst                          nz_linz_nzgd2000-c320110613-grid01.tif   nz_linz_nzgd2000-ka20161114-grid10.tif   nz_linz_stisht1977-nzvd2016.tif
nz_linz_auckht1946-nzvd2016.tif  nz_linz_nzgd2000-c420111223-grid01.tif   nz_linz_nzgd2000-ka20161114-grid11.tif   nz_linz_taraht1970-nzvd2016.tif
nz_linz_blufht1955-nzvd2016.tif  nz_linz_nzgd2000-ch20160214-grid01.tif   nz_linz_nzgd2000-ka20161114-grid12.tif   nz_linz_wellht1953-nzvd2016.tif
nz_linz_dublht1960-nzvd2016.tif  nz_linz_nzgd2000-cs20130721-grid01.tif   nz_linz_nzgd2000-lg20130816-grid01.tif   other.extra
nz_linz_duneht1958-nzvd2016.tif  nz_linz_nzgd2000-cs20130721-grid02.tif   nz_linz_nzgd2000-lg20130816-grid02.tif   proj.db
nz_linz_gisbht1926-nzvd2016.tif  nz_linz_nzgd2000-ds20090715-grid011.tif  nz_linz_nzgd2000-mq20041223-grid011.tif  proj.ini
nz_linz_lyttht1937-nzvd2016.tif  nz_linz_nzgd2000-ds20090715-grid012.tif  nz_linz_nzgd2000-mq20041223-grid012.tif  projjson.schema.json
nz_linz_motuht1953-nzvd2016.tif  nz_linz_nzgd2000-ds20090715-grid013.tif  nz_linz_nzgd2000-mq20041223-grid013.tif  README-PROJ-DATA.md
nz_linz_napiht1962-nzvd2016.tif  nz_linz_nzgd2000-ds20090715-grid014.tif  nz_linz_nzgd2000-mq20041223-grid014.tif  triangulation.schema.json
nz_linz_nelsht1955-nzvd2016.tif  nz_linz_nzgd2000-gs20071016-grid01.tif   nz_linz_nzgd2000-mq20041223-grid015.tif  world
nz_linz_nzgd2000-20000101.json   nz_linz_nzgd2000-ka20161114-grid01.tif   nz_linz_nzgd2000-mq20041223-grid016.tif
```

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/by-name/pr/proj-data/tests.nix [hunk 1]
```diff
@@ -0,0 +1,21 @@
+{ runCommand, proj, proj-data }:
+
+let
+  inherit (proj-data) pname;
+in
+runCommand "${pname}-tests" { meta.timeout = 60; }
+  ''
+    # add proj-data files to proj resources search path
+    export PROJ_DATA=${proj}/share/proj:${proj-data}
```
Comment (l0b0) [inline pkgs/by-name/pr/proj-data/tests.nix:9]: Should this possibly be set by the package itself?

File: pkgs/by-name/pr/proj-data/package.nix [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+    shopt -u extglob
+    runHook postInstall
+  '';
+
+  passthru.tests =
+    let
+      # build custom package containing `nz_linz` grids
+      projDataWithGrid = finalAttrs.finalPackage.overrideAttrs (_: {
+        postInstall = "cp nz_linz/* $out/";
```
Comment (l0b0) [inline pkgs/by-name/pr/proj-data/package.nix:47]: Shouldn't this override `gridPackages` instead?

File: pkgs/by-name/pr/proj-data/tests.nix [hunk 3]
```diff
@@ -0,0 +1,21 @@
+{ runCommand, proj, proj-data }:
+
+let
+  inherit (proj-data) pname;
+in
+runCommand "${pname}-tests" { meta.timeout = 60; }
+  ''
```
Comment (l0b0) [inline pkgs/by-name/pr/proj-data/tests.nix:7]: This test is using pipelines, so `set -o pipefail` might help.

File: pkgs/development/libraries/proj/default.nix [hunk 4]
```diff
@@ -58,6 +62,10 @@ stdenv.mkDerivation (finalAttrs: rec {
 
   doCheck = true;
 
+  postInstall = lib.optionalString withProjData ''
+    cp -R ${proj-data}/* $out/share/proj/
```
[the remark this answers is not in the document]
Comment (imincik) [reply pkgs/development/libraries/proj/default.nix:66]: > * GNU cp supports all of -R, -r, --recursive, and the last one is the most obvious.

Done. Thanks

> * If I understand correctly $out/share/proj doesn't exist at this point

Yes, it already exists and contains some files. No grid files should be hidden files, so I have no worries about not copying something what should be.

> * Quote variables for general safety.

I understand this point, but most of path variables in nixpkgs code (including the ones in this package) are not quoted. Nix store path can't contain any strange characters or spaces. I am leaving this unquoted for consistency.

File: pkgs/by-name/pr/proj-data/package.nix [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+    repo = "PROJ-data";
+    rev = finalAttrs.version;
+    hash = "sha256-/EgDeWy2+KdcI4DLsRfWi5OWcGwO3AieEJQ5Zh+wdYE=";
+  };
+
+  installPhase = ''
+    runHook preInstall
+    shopt -s extglob
+
+    mkdir -p $out
```
[the remark this answers is not in the document]
Comment (imincik) [reply pkgs/by-name/pr/proj-data/package.nix:27]: See the comment about store path variables quoting above.

File: pkgs/by-name/pr/proj-data/package.nix [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+    cp README.DATA $out/README-PROJ-DATA.md
```
[the remark this answers is not in the document]
Comment (imincik) [reply pkgs/by-name/pr/proj-data/package.nix:28]: See the comment about store path variables quoting above.

File: pkgs/by-name/pr/proj-data/package.nix [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+
+    for grid in ${builtins.toString gridPackages}; do
+      if [ ! -d $grid ]; then
```
[the remark this answers is not in the document]
Comment (imincik) [reply pkgs/by-name/pr/proj-data/package.nix:31]: If this check doesn't exist, the error message is quite confusing

```
       > Running phase: installPhase
       > cp: missing destination file operand after '/nix/store/v3258iwkrrmh5iddli7sdzxgikiw3cyh-proj-data-1.16.0/'
       > Try 'cp --help' for more information.
```.

I prefer to leave it as is.

File: pkgs/by-name/pr/proj-data/package.nix [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+        echo "ERROR: Grid ($grid) does not exist."
+        exit 1
+      fi
+
+      cp $grid/!(*.sh|*.py) $out/
```
[the remark this answers is not in the document]
Comment (imincik) [reply pkgs/by-name/pr/proj-data/package.nix:36]: Good point, but it is very hard to create future-proof list of allowed file types. I would prefer to leave it as is for now.

File: pkgs/by-name/pr/proj-data/tests.nix (hunk 1 above)
Comment (imincik) [reply pkgs/by-name/pr/proj-data/tests.nix:9]: I am not sure what is the best approach. We can wrap all executables with `PROJ_DATA` path, but it wouldn't help with dozens of other apps using `proj` as library (like QGIS for example).

File: pkgs/by-name/pr/proj-data/package.nix (hunk 2 above)
Comment (imincik) [reply pkgs/by-name/pr/proj-data/package.nix:47]: Yes, I would like to, but `finalPackage` doesn't support `override` (it supports only `overrideAttrs`). Any suggestions are welcome.

File: pkgs/by-name/pr/proj-data/tests.nix (hunk 3 above)
Comment (imincik) [reply pkgs/by-name/pr/proj-data/tests.nix:7]: Good point, done.

**Comment by imincik** [comment]
@dotlambda , are you OK with this PR ?

File: pkgs/development/libraries/proj/default.nix (hunk 4 above)
[the remark this answers is not in the document]
Comment (l0b0) [reply pkgs/development/libraries/proj/default.nix]: (Feel free to close any discussions where I've :+1:'ed your last reply - I don't have access to close discussions.)

File: pkgs/by-name/pr/proj-data/package.nix (hunk 7 above)
[the remark this answers is not in the document]
Comment (l0b0) [reply pkgs/by-name/pr/proj-data/package.nix:31]: Sounds good. In that case I'd recommend changing the check to `if ! [[ -d "$grid" ]]` - it won't produce a syntax error if anyone includes an empty grid package by accident.

File: pkgs/by-name/pr/proj-data/tests.nix (hunk 1 above)
Comment (l0b0) [reply pkgs/by-name/pr/proj-data/tests.nix:11]: > I am not sure what is the best approach. We can wrap all executables with `PROJ_DATA` path, […]

Sounds good.

> but it wouldn't help with dozens of other apps using `proj` as library (like QGIS for example).

Isn't that unrelated? We'd be making one thing work without breaking anything else, right?

File: pkgs/by-name/pr/proj-data/package.nix (hunk 2 above)
Comment (l0b0) [reply pkgs/by-name/pr/proj-data/package.nix:47]: I don't know Nix well enough to suggest anything. Feel free to ignore this, unless someone else has a suggestion.

**Review by l0b0 — CHANGES_REQUESTED** [review]
https://github.com/NixOS/nixpkgs/pull/280062/files#r1458004097 is still applicable.

**Comment by imincik** [comment]
Consider using `proj.[passthru.]withProjData` instead of function parameter.

**Comment by imincik** [comment]
I am replacing this PR with #290643 which is using better approach of adding grid packages to `proj-data` using `withGridPackages` passthru.

**Pull request closed without merging**